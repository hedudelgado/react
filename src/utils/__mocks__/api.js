/* eslint-disable */
export const getCountries = jest.fn(() => Promise.resolve([
  {
    countryCode: 'GB',
    countryCodeISO: 'GB',
    countryLegend: 'United Kingdom (the)',
    passportRequired: false,
    dialingCode: '+44',
    flagImg: '/content/dam/global/flags/United-Kingdom.png',
  },
  {
    countryCode: 'D',
    countryCodeISO: 'DE',
    countryLegend: 'Germany',
    passportRequired: true,
    dialingCode: '+49',
    flagImg: '/content/dam/global/flags/Germany.png',
  },
  {
    countryCode: 'GR',
    countryCodeISO: 'GR',
    countryLegend: 'Greece',
    passportRequired: true,
    dialingCode: '+30',
    flagImg: '/content/dam/global/flags/Greece.png',
  },
]));

export const getDictionary = jest.fn(() => Promise.resolve({
  'profile.button.changepassword': 'Change password',
  'profile.button.edit': 'Edit Profile',
  'profile.preview.password': 'Password ************',
  'profile.title': 'Your Profile',
}));
