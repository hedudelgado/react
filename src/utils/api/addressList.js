import fetch from './utils/fetchWrapper';
import { getJsonItem } from '../session-storage';

const extractAddresses = ({ addresses }) => addresses;

export const getAddressList = postCode => fetch(`${getJsonItem('environment').apiURL}/addresses/${postCode}`)
  .then(extractAddresses);
