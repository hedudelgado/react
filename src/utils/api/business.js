import fetch from './utils/fetchWrapper';
import { getJsonItem } from '../session-storage';

export const tetherLogin = guid => fetch(`${getJsonItem('environment').apiURL}/business/tether/login`,
  {
    method: 'POST',
    body: JSON.stringify({ guid }),
  });

export const resetMemorableWord = payload => fetch(`${getJsonItem('environment').apiURL}/auth/hotels/memorable-word/reset`,
  {
    method: 'POST',
    body: JSON.stringify(payload),
  });
