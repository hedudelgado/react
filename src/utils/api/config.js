import fetch from './utils/fetchWrapper';

export const getConfig = configUrl => fetch(configUrl);
