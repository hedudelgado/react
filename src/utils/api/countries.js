import fetch from './utils/fetchWrapper';
import { getJsonItem } from '../session-storage';

const extractCountries = ({ countries }) => countries;

const sortByCountryLegend = (a, b) => ((a.countryLegend > b.countryLegend) ? 1 : -1);
const sortCountries = countries => countries.sort(sortByCountryLegend);

export const getCountries = () => fetch(`${getJsonItem('environment').apiURL}/countries`)
  .then(extractCountries)
  .then(sortCountries);
