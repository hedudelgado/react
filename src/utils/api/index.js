export {
  getConfig,
} from './config';

export {
  getCountries,
} from './countries';

export {
  getNewsletterRegions,
  getNewsletterPreferences,
  updateNewsletter,
} from './newsletter';

export {
  updateProfile,
  deleteProfile as deleteCustomer,
  getProfile,
} from './profile';

export {
  getAddressList,
} from './addressList';

export {
  default as validateBin,
} from './validateBin';

export {
  resetMemorableWord,
  tetherLogin,
} from './business';
