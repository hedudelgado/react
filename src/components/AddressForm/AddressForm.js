import React, { Fragment, Component } from 'react';
import { object, array, func } from 'prop-types';
import Input from '../../layout/Input';
import Select from '../../layout/Select';
import Button from '../../layout/Button';
import Typography from '../../layout/Typography';
import RadioGroup, { RadioItem } from '../../layout/RadioGroup';
import { getAddressList } from '../../utils/api';
import { selectedFlagImage } from '../../utils/selectedFlagImage';
import { BUSINESS, LEISURE } from '../../config.json';

class AddressForm extends Component {
  constructor(props) {
    super(props);

    const countryOptions = props.countries.map(country => (
      <option
        label={country.countryLegend}
        value={country.countryCode}
        key={country.countryCode}
      >
        {country.countryLegend}
      </option>
    ));

    this.state = {
      countryOptions,
      addressList: [],
      addressListOptions: [],
      disableSelectDefault: false,
    };

    this.handleAddressChange = this.handleAddressChange.bind(this);
    this.handleAddressList = this.handleAddressList.bind(this);
  }

  handleAddressChange(event) {
    if (event.target.value !== 'default') {
      const {
        setFieldValue,
      } = this.props;
      const { addressList } = this.state;
      const address = addressList[event.target.value];

      this.setState({
        disableSelectDefault: true,
      });

      ['line1', 'line2', 'line3', 'line4'].forEach((field) => {
        setFieldValue(field, address[field] ? address[field] : '');
      });

      setFieldValue('postCode', address.postcode);
      setFieldValue('countryCode', address.country);
      setFieldValue('companyName', address.companyName ? address.companyName : '');
      setFieldValue('type', address.companyName ? BUSINESS : LEISURE);
    }
  }

  handleAddressList(postcode) {
    getAddressList(postcode).then((response) => {
      this.setState({
        addressList: response,
        addressListOptions: response.map((address, key) => (
          <option
            label={address.label}
            value={key}
            key={address.label}
          >
            {address.label}
          </option>
        )),
        disableSelectDefault: false,
      });
    });
  }

  render() {
    const {
      dictionary,
      values: {
        postCode,
        type,
        countryCode,
        line1,
        line2,
        line3,
        line4,
        companyName,
      },
      errors,
      touched,
      handleBlur,
      handleChange,
      countries,
    } = this.props;

    const {
      countryOptions,
      addressList,
      addressListOptions,
      disableSelectDefault,
    } = this.state;

    return (
      <Fragment>
        <Typography component="h6" className="pi-color-greyscale-darkest mt0 mb1">
          {dictionary['profile.postCodeTitle']}
        </Typography>
        <Input
          type="text"
          name="postCode"
          id="postCode"
          value={postCode}
          placeholder={dictionary['profile.form.postcodeuk']}
          onChange={handleChange}
          onBlur={handleBlur}
          error={dictionary[errors.postCode]}
          touched={touched.postCode}
        />
        <Button
          data-test="find-postcode"
          color="secondary"
          onClick={() => this.handleAddressList(postCode)}
          className="wb-push--bottom wb-btn--full-width"
        >
          {dictionary['profile.button.findpostcode']}
        </Button>
        {!!addressList.length && (
          <Select
            name="addressList"
            id="addressList"
            onBlur={handleBlur}
            onChange={this.handleAddressChange}
            sessioncamhide
          >
            <option key="default" value="default" defaultValue disabled={disableSelectDefault}>
              {dictionary['profile.placeholder.address']}
            </option>
            {addressListOptions}
          </Select>
        )}
        <RadioGroup>
          <RadioItem
            field={{
              name: 'type',
              value: type,
              onBlur: handleBlur,
              onChange: handleChange,
            }}
            label={dictionary['profile.radio.address.home']}
            name="type"
            id="homeAddress"
            value={LEISURE}
            sessioncamhide
          />
          <RadioItem
            field={{
              name: 'type',
              value: type,
              onBlur: handleBlur,
              onChange: handleChange,
            }}
            label={dictionary['profile.radio.address.business']}
            name="type"
            id="businessAddress"
            value={BUSINESS}
            sessioncamhide
          />
        </RadioGroup>
        {type === BUSINESS && (
          <Input
            type="text"
            name="companyName"
            id="companyName"
            value={companyName}
            placeholder={dictionary['profile.form.companyname']}
            onChange={handleChange}
            onBlur={handleBlur}
            error={dictionary[errors.companyName]}
            touched={touched.companyName}
          />
        )}
        <Input
          type="text"
          name="line1"
          id="line1"
          value={line1}
          placeholder={dictionary['profile.form.address.line1']}
          onChange={handleChange}
          onBlur={handleBlur}
          error={dictionary[errors.line1]}
          touched={touched.line1}
          required
        />
        <Input
          type="text"
          name="line2"
          id="line2"
          value={line2}
          placeholder={dictionary['profile.form.address.line2']}
          onChange={handleChange}
          onBlur={handleBlur}
          error={dictionary[errors.line2]}
          touched={touched.line2}
        />
        <Input
          type="text"
          name="line3"
          id="line3"
          value={line3}
          placeholder={dictionary['profile.form.address.line3']}
          onChange={handleChange}
          onBlur={handleBlur}
          error={dictionary[errors.line3]}
          touched={touched.line3}
        />
        <Input
          type="text"
          name="line4"
          id="line4"
          value={line4}
          placeholder={dictionary['profile.form.address.line4']}
          onChange={handleChange}
          onBlur={handleBlur}
          error={dictionary[errors.line4]}
          touched={touched.line4}
        />
        <Input
          type="text"
          name="postCode"
          id="postCode"
          value={postCode}
          placeholder={dictionary['profile.form.postcode']}
          onChange={handleChange}
          onBlur={handleBlur}
          error={dictionary[errors.postCode]}
          touched={touched.postCode}
        />
        <Select
          onBlur={handleBlur}
          onChange={handleChange}
          value={countryCode}
          name="countryCode"
          id="countryCode"
          flagImg={selectedFlagImage(countryCode, countries)}
          touched={touched.countryCode}
          sessioncamhide
        >
          {countryOptions}
        </Select>
      </Fragment>
    );
  }
}

AddressForm.propTypes = {
  dictionary: object.isRequired,
  countries: array.isRequired,
  values: object.isRequired,
  setFieldValue: func.isRequired,
  handleBlur: func.isRequired,
  handleChange: func.isRequired,
  errors: object.isRequired,
  touched: object.isRequired,
};

export default AddressForm;
