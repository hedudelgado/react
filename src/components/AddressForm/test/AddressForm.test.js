import React from 'react';
import { shallow } from 'enzyme';

import AddressForm from '../AddressForm';
import { getAddressList } from '../../../utils/api';

jest.mock('../../../utils/api', () => ({
  __esModule: true,
  getAddressList: jest.fn(() => Promise.resolve([
    {
      postcode: 'EC1N 2TD',
      label: 'Brainbow Ltd, 120 Holborn, LONDON, EC1N 2TD',
      line1: '120 Holborn',
      line4: 'LONDON',
      country: 'GB',
      companyName: 'Brainbow Ltd',
    },
    {
      postcode: 'EC1N 2TD',
      label: 'North Highland, 120 Holborn, LONDON, EC1N 2TD',
      line1: '120 Holborn',
      line4: 'LONDON',
      country: 'GB',
      companyName: 'North Highland',
    },
    {
      postcode: 'EC1N 2TD',
      label: 'North Highland UK (Holdings) Ltd, 120 Holborn, LONDON, EC1N 2TD',
      line1: '120 Holborn',
      line4: 'LONDON',
      country: 'GB',
      companyName: 'North Highland UK (Holdings) Ltd',
    },
    {
      postcode: 'NW2 3AN',
      label: '27a Chichele Road, LONDON, NW2 3AN',
      line1: '27a Chichele Road',
      line4: 'LONDON',
      country: 'GB',
    },
  ])),
}));


describe('AddressForm Component', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      selectedImage: jest.fn(),
      setFieldValue: jest.fn(),
      handleBlur: jest.fn(),
      handleChange: jest.fn(),
      dictionary: {
        'profile.radio.address.home': 'Home address',
        'profile.radio.address.business': 'Business address',
        'profile.button.findpostcode': 'Find new address',
      },
      errors: {},
      touched: {},
      values: {
        countryCode: 'GB',
        type: 'HOME',
      },
      countries: [
        {
          countryCode: 'GB',
          countryCodeISO: 'GB',
          countryLegend: 'United Kingdom (the)',
          passportRequired: false,
          dialingCode: '+44',
          flagImg: '/content/dam/global/flags/United-Kingdom.png',
        },
        {
          countryCode: 'D',
          countryCodeISO: 'DE',
          countryLegend: 'Germany',
          passportRequired: true,
          dialingCode: '+49',
          flagImg: '/content/dam/global/flags/Germany.png',
        },
        {
          countryCode: 'GR',
          countryCodeISO: 'GR',
          countryLegend: 'Greece',
          passportRequired: true,
          dialingCode: '+30',
          flagImg: '/content/dam/global/flags/Greece.png',
        },
        {
          countryCode: 'TEST',
          countryCodeISO: 'TEST',
          countryLegend: 'Test',
          passportRequired: true,
          dialingCode: '+499',
          flagImg: 'test.png',
        },
      ],
    };
    wrapper = shallow(<AddressForm {...props} />);
  });

  describe('Renders AddressForm Element', () => {
    it('should render the component', () => {
      expect(wrapper).not.toBeEmptyRender();
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should display the two address radio buttons', () => {
      expect(wrapper.find('RadioGroup [id$="Address"]')).toHaveLength(2);
    });

    it('should not display the company name field for home addresses', () => {
      expect(wrapper.find('[name="companyName"]')).toHaveLength(0);
    });

    it('should display the company name field for business addresses', () => {
      const type = 'BUSINESS';
      const businessProps = {
        ...props,
        values: {
          ...props.values,
          type,
        },
      };

      wrapper = shallow(<AddressForm {...businessProps} />);
      expect(wrapper.find('[name="companyName"]')).toHaveLength(1);
    });
  });

  describe('Gets list of addresses and updates state', () => {
    it('should not display addressList when addressList is empty', () => {
      expect(wrapper.find('#addressList')).toHaveLength(0);
    });

    it('should display addressList', async () => {
      wrapper.find('[data-test="find-postcode"]').simulate('click');
      await Promise.all([getAddressList]).then(() => {
        expect(wrapper.state().addressList).toHaveLength(4);
        expect(wrapper.state().addressListOptions).toHaveLength(4);
      });
    });

    it('should display addressList when addressList is empty', async () => {
      wrapper.find('[data-test="find-postcode"]').simulate('click');
      await Promise.all([getAddressList]).then(() => {
        expect(wrapper.find('#addressList'))
          .toHaveLength(1);
      });
    });

    it('should update addressFields when user selects a company address from dropdown', async () => {
      const event = { target: { value: '1' } };
      wrapper.find('[data-test="find-postcode"]').simulate('click');
      await Promise.all([getAddressList]).then(() => {
        wrapper.find('#addressList').simulate('change', event);
        expect(props.setFieldValue)
          .toHaveBeenNthCalledWith(1, 'line1', '120 Holborn')
          .toHaveBeenNthCalledWith(2, 'line2', '')
          .toHaveBeenNthCalledWith(3, 'line3', '')
          .toHaveBeenNthCalledWith(4, 'line4', 'LONDON')
          .toHaveBeenNthCalledWith(5, 'postCode', 'EC1N 2TD')
          .toHaveBeenNthCalledWith(6, 'countryCode', 'GB')
          .toHaveBeenNthCalledWith(7, 'companyName', 'North Highland');
      });
    });

    it('should update addressFields when user selects a house address from dropdown', async () => {
      const event = { target: { value: '3' } };
      wrapper.find('[data-test="find-postcode"]').simulate('click');
      await Promise.all([getAddressList]).then(() => {
        wrapper.find('#addressList').simulate('change', event);
        expect(props.setFieldValue)
          .toHaveBeenNthCalledWith(1, 'line1', '27a Chichele Road')
          .toHaveBeenNthCalledWith(2, 'line2', '')
          .toHaveBeenNthCalledWith(3, 'line3', '')
          .toHaveBeenNthCalledWith(4, 'line4', 'LONDON')
          .toHaveBeenNthCalledWith(5, 'postCode', 'NW2 3AN')
          .toHaveBeenNthCalledWith(6, 'countryCode', 'GB')
          .toHaveBeenNthCalledWith(7, 'companyName', '');
      });
    });
  });

  describe('Manages the Address Select placeholder', () => {
    it('should stop the execution if the default placeholder is somehow selected', () => {
      const event = { target: { value: 'default' } };
      expect(wrapper.instance().handleAddressChange(event)).toBe(undefined);
    });

    it('should disable the placeholder option if a real selection has been made', async () => {
      const event = { target: { value: '3' } };
      wrapper.find('[data-test="find-postcode"]').simulate('click');
      await Promise.all([getAddressList]).then(() => {
        wrapper.find('#addressList').simulate('change', event);
        expect(wrapper.state().disableSelectDefault).toBe(true);
      });
    });
  });
});
