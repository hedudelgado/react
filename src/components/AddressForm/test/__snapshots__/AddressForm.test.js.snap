// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressForm Component Renders AddressForm Element should match snapshot 1`] = `
<Fragment>
  <Typography
    className="pi-color-greyscale-darkest mt0 mb1"
    component="h6"
  />
  <Input
    id="postCode"
    mb="wb-push-middle--bottom"
    name="postCode"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="text"
  />
  <Button
    className="wb-push--bottom wb-btn--full-width"
    color="secondary"
    data-test="find-postcode"
    fullWidth={false}
    href={null}
    onClick={[Function]}
    submit={false}
    variant="button"
  >
    Find new address
  </Button>
  <RadioGroup>
    <RadioItem
      field={
        Object {
          "name": "type",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "HOME",
        }
      }
      id="homeAddress"
      label="Home address"
      name="type"
      sessioncamhide={true}
      subLabels={Array []}
      value="HOME"
    />
    <RadioItem
      field={
        Object {
          "name": "type",
          "onBlur": [MockFunction],
          "onChange": [MockFunction],
          "value": "HOME",
        }
      }
      id="businessAddress"
      label="Business address"
      name="type"
      sessioncamhide={true}
      subLabels={Array []}
      value="BUSINESS"
    />
  </RadioGroup>
  <Input
    id="line1"
    mb="wb-push-middle--bottom"
    name="line1"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    required={true}
    sessioncamhide={true}
    type="text"
  />
  <Input
    id="line2"
    mb="wb-push-middle--bottom"
    name="line2"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="text"
  />
  <Input
    id="line3"
    mb="wb-push-middle--bottom"
    name="line3"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="text"
  />
  <Input
    id="line4"
    mb="wb-push-middle--bottom"
    name="line4"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="text"
  />
  <Input
    id="postCode"
    mb="wb-push-middle--bottom"
    name="postCode"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    sessioncamhide={true}
    type="text"
  />
  <Select
    flagImg="/content/dam/global/flags/United-Kingdom.png"
    halfWidth={false}
    hideIcon={false}
    hideMessage={false}
    id="countryCode"
    name="countryCode"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    placeholder=""
    sessioncamhide={true}
    value="GB"
  >
    <option
      key="GB"
      label="United Kingdom (the)"
      value="GB"
    >
      United Kingdom (the)
    </option>
    <option
      key="D"
      label="Germany"
      value="D"
    >
      Germany
    </option>
    <option
      key="GR"
      label="Greece"
      value="GR"
    >
      Greece
    </option>
    <option
      key="TEST"
      label="Test"
      value="TEST"
    >
      Test
    </option>
  </Select>
</Fragment>
`;
